### ISSUE TITLE
ConstantLengthDataset does not return the right length 

### ISSUE THREAD

**Issue by edbeeching** [huggingface/trl#943]
We noticed that when training with longer sequence lengths and packing=True that the estimated steps for an epoch can be far lower than expected.  cc @lewtun 
Example:
![image](https://github.com/huggingface/trl/assets/7275864/54ac4b39-33fd-462c-8881-4f17b4d44225)
The root cause of this appears to be how the length of the `ConstantLengthDataset` is calculated, currently it returns the length of the unpacked dataset. https://github.com/huggingface/trl/blob/5b32372b71c5699dbb6d83b9a9c8e01ae45eb6c0/trl/trainer/utils.py#L554

Minimal example:
```python
from datasets import load_dataset
from transformers import AutoTokenizer
from trl.trainer.utils import ConstantLengthDataset

tokenizer = AutoTokenizer.from_pretrained("EleutherAI/pythia-70m")
dataset = load_dataset("timdettmers/openassistant-guanaco", split="train")

packed_dataset = ConstantLengthDataset(
    tokenizer=tokenizer,
    dataset=dataset,
    dataset_text_field="text",
    num_of_sequences=10000,
)

length = len(packed_dataset)

actual_length = 0

for example in packed_dataset:
    actual_length += 1
    
print(f"{length=}")
print(f"{actual_length=}")
print(f"{actual_length==length=}")
```
```
length=9846
actual_length=4283
actual_length==length=False
```

## Potential problems this causes
This may lead to the warmup steps and other step related options (linear, cosine) to be calculated incorrectly.

## Potential Solution
Perform the packing upfront in  `__init__` method, return the len of the packed examples. Modify the `__iter__` method to return the precomputed packed sequences. This may cause issues with large datasets, small buffers and the infinite option.

**Comment by vwxyzjn** [comment]
Thanks @edbeeching for the report! I think this is really a `seq_length` problem: 

```
from datasets import load_dataset
from transformers import AutoTokenizer
from trl.trainer.utils import ConstantLengthDataset

tokenizer = AutoTokenizer.from_pretrained("EleutherAI/pythia-70m")
dataset = load_dataset("timdettmers/openassistant-guanaco", split="train")
print(f"{len(dataset)=}")    

for seq_length in [512, 1024, 2048]:
    packed_dataset = ConstantLengthDataset(
        tokenizer=tokenizer,
        dataset=dataset,
        dataset_text_field="text",
        # num_of_sequences=10000,
        seq_length=seq_length,
    )
    actual_length = 0
    for example in packed_dataset:
        actual_length += 1
    print(f"{seq_length=}, {actual_length=}")

len(dataset)=9846
seq_length=512, actual_length=8563
seq_length=1024, actual_length=4281
seq_length=2048, actual_length=2141
```


The larger the `seq_length` the more examples we can "pack" into a single sequence, the less 'packed_dataset''s actual lengths.

**Comment by lvwerra** [comment]
Yes, since we pack the dataset on the fly, we don't know in advance how many samples it will yield. We could add a `precompute` flag and run the processing in advance. This should be fine for smaller datasets for sure.

**Comment by tcapelle** [comment]
I am also curious on how the Zephyr fine tune works as it uses Cosine Scheduler and packing, so the total training steps is not know in advance.

**Comment by alvarobartt** [comment]
Hi @edbeeching, I'm struggling a bit to reproduce Zephyr when trying to push the model to the Hub on every epoch, as it seems that the epoch logging is messed up due to this issue, and then it's crashing when trying to push to the Hub, could this be related? I'm using a similar configuration to the one provided at https://github.com/huggingface/alignment-handbook/blob/main/recipes/zephyr-7b-beta/sft/config_full.yaml, but adding `save_strategy: "epoch"` instead of `save_strategy: "none"`. I'm running it with `accelerate launch` and using also the `deepspeed_zero3.yaml` configuration provided there, but pushing to the Hub seems to explore the run, could this issue be related? I'm trying now to manually set the steps and save via `save_steps`

**Comment by maneandrea** [comment]
Probably not terribly useful, but I have encountered the same problem in the case where the input dataset consists of strings longer than the desired `seq_length` (so that more examples are yielded by a single entry in `self.dataset`). In this case I get a lot of warnings of the form:
```
UserWarning: Length of IterableDataset <trl.trainer.utils.ConstantLengthDataset object at 0x7f873937a950> was reported to be 1 (when accessing len(dataloader)), but 3 samples have been fetched.
UserWarning: Length of IterableDataset <trl.trainer.utils.ConstantLengthDataset object at 0x7f873937a950> was reported to be 1 (when accessing len(dataloader)), but 4 samples have been fetched.
...
```
(the numbers are so low because this was a very small example for debugging)

**Comment by alvarobartt** [comment]
Sharing this here in case it's useful for anyone 🤗 (forgot to share before)

https://gist.github.com/alvarobartt/d08888dd2660b6763421dd6b1142127c

**Comment by lvwerra** [comment]
I think these issues should have been fixed in #979 - now the packed dataset is precomputed and the length/epoch should match what is provided at the cost of a small overhead at the beginning to process the dataset.
